Remove native_functions.yaml dependency from Activation.cu

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64499

This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

peterbell10 committed:
Remove native_functions.yaml dependency from Activation.cu

This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the number
random generation code, so I've moved it into its own file.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

[ghstack-poisoned]

ezyang committed:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30867323).

ezyang commented on aten/src/ATen/native/cuda/Activation.cu:
```diff
@@ -209,18 +144,14 @@ void log_sigmoid_backward_kernel(TensorIterator& iter) {
+        [weight_data] GPU_LAMBDA (scalar_t input_val) {
+          return (input_val > 0) ? input_val : *weight_data * input_val;
+        });
+  });
```
ugh, all of the code movement makes the diff unreadable

peterbell10 replied:
The diff ignoring white space changes makes these `launch_` functions clearer:
https://github.com/pytorch/pytorch/pull/64499/files?diff=unified&w=1

The only substantive change is dispatch now uses `iter.input_dtype()` instead of `input.scalar_type()`. 

Everything else is just moving code around. You can see this by diffing the old `Activation.cu` with `Activation.cpp` and looking at what's added by grepping for lines that start with `>` indicating added

```sh
$ git checkout HEAD~1 -- ./Activation.cu
$ diff Activation.cu Activation.cpp -w | grep "^>"
> #include <ATen/native/cuda/Activation.h>
> namespace at { namespace native {
>   launch_log_sigmoid_forward_kernel(iter);
>     auto iter = TensorIterator::unary_op(result, input);
>     launch_prelu_cuda_kernel_share_weights(iter, weight);
>     launch_prelu_cuda_kernel_multi_weights(result, input, weight);
>     at::TensorIterator iter = TensorIteratorConfig()
>         .add_output(input_grad)
>         .add_output(weight_grad_collector)
>         .add_input(input)
>         .add_input(grad_out)
>         .build();
> 
>     launch_prelu_cuda_backward_kernel_share_weights(iter, weight);
>     launch_prelu_cuda_backward_kernel_multi_weights(
>         input, weight, grad_out, input_grad, weight_grad_collector);
> }}  // namespace at::native
```

All of these are:
- Includes & namespace stuff
- calling `launch_`... functions instead of embedding the kernel definition
- Moving `TensorIterator` construction out of the `_impl` functions, so we can pass them as arguments to `launch_`

And the same test works for `RreluWithNoise.cu`:
```sh
$ diff Activation.cu RreluWithNoise.cu -w | grep "^>"
> namespace at { namespace native {
> }}  // namespace at::native
```

ezyang commented on aten/src/ATen/native/cuda/Activation.cu:
```diff
@@ -241,91 +172,60 @@ __global__ void prelu_cuda_kernel_multi_weights(
-
-  TORCH_CHECK(input.is_contiguous());
-  TORCH_CHECK(weight.is_contiguous());
+void launch_prelu_cuda_kernel_multi_weights(
```
this is `// case2: multiple parameters, one for each channel`

ezyang reviewed (approved):
this approve predicated on most things being code being moved around, no substantive changes

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30867323).

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

ezyang committed:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30867323).

Commit:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30867323).

Commit:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30867323).

Commit:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30867323).

Commit:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from Activation.cu"


This moves the native functions into a separate Activation.cpp file,
which calls into `launch_..._kernel` functions defined in `Activation.cu`.
The exception is `rrelu_with_noise` which is compilcated by the 
random number generation code, so I've moved it into its own file.

Differential Revision: [D30867323](https://our.internmc.facebook.com/intern/diff/D30867323)

[ghstack-poisoned]
